runtime/runtime_controller.cc

```diff
@@ -44,47 +46,62 @@ RuntimeController::RuntimeController(
-  auto strong_root_isolate =
-      DartIsolate::CreateRootIsolate(vm_->GetVMData()->GetSettings(),  //
-                                     isolate_snapshot_,                //
-                                     task_runners_,                    //
-                                     std::make_unique<Window>(this),   //
-                                     snapshot_delegate_,               //
-                                     io_manager_,                      //
-                                     unref_queue_,                     //
-                                     image_decoder_,                   //
-                                     p_advisory_script_uri,            //
-                                     p_advisory_script_entrypoint,     //
-                                     nullptr,                          //
-                                     isolate_create_callback_,         //
-                                     isolate_shutdown_callback_        //
-                                     )
-          .lock();
-
-  FML_CHECK(strong_root_isolate) << "Could not create root isolate.";
-
-  // The root isolate ivar is weak.
-  root_isolate_ = strong_root_isolate;
-
-  strong_root_isolate->SetReturnCodeCallback([this](uint32_t code) {
-    root_isolate_return_code_ = {true, code};
-  });
-
-  if (auto* window = GetWindowIfAvailable()) {
-    tonic::DartState::Scope scope(strong_root_isolate);
-    window->DidCreateIsolate();
-    if (!FlushRuntimeStateToIsolate()) {
-      FML_DLOG(ERROR) << "Could not setup initial isolate state.";
-    }
-  } else {
-    FML_DCHECK(false) << "RuntimeController created without window binding.";
-  }
-
-  FML_DCHECK(Dart_CurrentIsolate() == nullptr);
+  fml::MessageLoop::GetCurrent().GetTaskRunner()->PostTask(
+      [self = weak_factory_.GetWeakPtr()]() {
+        if (!self) {
+          return;
+        }
+
+        auto strong_root_isolate =
+            DartIsolate::CreateRootIsolate(
+                self->vm_->GetVMData()->GetSettings(),  //
+                self->isolate_snapshot_,                //
+                self->task_runners_,                    //
+                std::make_unique<Window>(self.get()),   //
+                self->snapshot_delegate_,               //
+                self->io_manager_,                      //
+                self->unref_queue_,                     //
+                self->image_decoder_,                   //
+                self->advisory_script_uri_,             //
+                self->advisory_script_entrypoint_,      //
+                nullptr,                                //
+                self->isolate_create_callback_,         //
+                self->isolate_shutdown_callback_        //
+                )
+                .lock();
+        FML_CHECK(strong_root_isolate) << "Could not create root isolate.";
+
+        // The root isolate ivar is weak.
+        self->root_isolate_ = strong_root_isolate;
```

Comment:
The API as originally designed expects the root isolate to be ready when the runtime controller has been created (this assumption is prevalent throughout the engine). Now, a race has been introduced with no way of telling from the caller about when the root isolate is ready (or the other ivars).

Reply:
Yes, it breaks the assumption but also makes sure the root isolate must be ready before running.